packages/app/src/cli/services/dev.ts

```diff
@@ -65,6 +66,11 @@ interface DevWebOptions {
 }
 
 async function dev(options: DevOptions) {
+  // Be optimistic about tunnel creation and do it as early as possible
```

Comment:
The scattering of tunnel-related logic in the function makes it hard to reason about and brittle. What about having a function like:

```js
await startTunnelOptimistically({defaultTunnel: 'cloudflare', fallback: 'ngrok'}, async (tunnelReady) => {
 // First part of the dev logic
 
 const provider = await tunnelReady 

// Continue with dev
 
})
```

Reply:
Is not that straightforward, we don't use ngrok as a fallback, only if there is a flag enabled remotely that tell us to use it. This is the process:

1. Start cloudflare as early as possible
2. Authenticate User if needed
3. Fetch Org/App information
4. Check if the Org beta flags forces us to us ngrok (we added this as a safety measure)
5. If the flag tell us to use ngrok, stop cloudflare tunnel and start ngrok. If not, do nothing.
6. Later in the flow, request the tunnel URL for whatever tunnel provider is the active one.

If everything goes well, in a couple of releases we should completely remove ngrok, removing steps 4 and 5 which will simplify the tunneling logic in `dev`

Reply:
i've simplified this a bit. We now receive a tunnelClient that we pass to the needed part of the flow, we can ask the client to stop the tunnel or to get the current status